banyand/index/index.go

```diff
@@ -19,36 +19,230 @@ package index
+	"go.uber.org/multierr"
 
 	"github.com/apache/skywalking-banyandb/api/common"
+	"github.com/apache/skywalking-banyandb/api/event"
 	apiv1 "github.com/apache/skywalking-banyandb/api/proto/banyandb/v1"
 	"github.com/apache/skywalking-banyandb/banyand/discovery"
-	"github.com/apache/skywalking-banyandb/banyand/queue"
+	"github.com/apache/skywalking-banyandb/banyand/index/tsdb"
+	"github.com/apache/skywalking-banyandb/pkg/bus"
+	"github.com/apache/skywalking-banyandb/pkg/logger"
 	"github.com/apache/skywalking-banyandb/pkg/posting"
 	"github.com/apache/skywalking-banyandb/pkg/run"
 )
 
+var (
+	ErrShardNotFound       = errors.New("series doesn't exist")
+	ErrTraceSeriesNotFound = errors.New("trace series not found")
+	ErrUnknownField        = errors.New("the field is unknown")
+)
+
 type Condition struct {
 	Key    string
 	Values [][]byte
 	Op     apiv1.PairQuery_BinaryOp
 }
 
-//go:generate mockgen -destination=./index_mock.go -package=index . Repo
+type Field struct {
+	ChunkID common.ChunkID
+	Name    string
+	Value   []byte
+}
+
+//go:generate mockgen -destination=./index_mock.go -package=index . Service
 type Repo interface {
 	Search(seriesMeta common.Metadata, shardID uint, startTime, endTime uint64, indexObjectName string, conditions []Condition) (posting.List, error)
+	Insert(seriesMeta common.Metadata, shardID uint, fields *Field) error
 }
 
 type Builder interface {
-	run.Config
 	run.PreRunner
+	run.Service
 }
 
 type Service interface {
 	Repo
 	Builder
 }
 
-func NewService(ctx context.Context, repo discovery.ServiceRepo, pipeline queue.Queue) (Service, error) {
-	return nil, nil
+type series struct {
+	repo map[uint]*shard
+}
+
+type shard struct {
+	meta  map[string][]*apiv1.IndexObject
+	store tsdb.GlobalStore
+}
+
+type service struct {
+	meta              *indexMeta
+	log               *logger.Logger
+	repo              discovery.ServiceRepo
+	stopCh            chan struct{}
+	indexRuleListener *indexRuleListener
+}
+
+func NewService(_ context.Context, repo discovery.ServiceRepo) (Service, error) {
+	svc := &service{
+		repo:              repo,
+		indexRuleListener: &indexRuleListener{},
+	}
+	svc.meta = &indexMeta{
+		meta: make(map[string]*series),
+	}
+	svc.indexRuleListener.indexMeta = svc.meta
+	svc.indexRuleListener.closeFunc = func() {
+		svc.stopCh <- struct{}{}
+	}
+	return svc, nil
+}
+
+func (s *service) Insert(series common.Metadata, shardID uint, field *Field) error {
+	sd, err := s.getShard(series, shardID)
+	if err != nil {
+		return err
+	}
+	objects, ok := sd.meta[field.Name]
+	if !ok {
+		return ErrUnknownField
+	}
+	for _, object := range objects {
+		err = multierr.Append(err, sd.store.Insert(&tsdb.Field{
+			Name:  []byte(compositeFieldID(object.GetName(), field.Name)),
+			Value: field.Value,
+		}, field.ChunkID))
+	}
+	return err
+}
+
+func (s *service) getShard(series common.Metadata, shardID uint) (*shard, error) {
+	id := compositeSeriesID(series.Spec)
+	ss, ok := s.meta.meta[id]
```

Comment:
use `s.get(series *apiv1.Metadata)` in order to make use of read lock?

Reply:
done